src/django_components/slots.py

```diff
@@ -96,56 +109,83 @@ def __repr__(self) -> str:
         return f"<Slot Node: {self.name}. Contents: {repr(self.nodelist)}. Options: {self.active_flags}>"
 
     def render(self, context: Context) -> SafeString:
-        try:
-            filled_slots_map: FilledSlotsContext = context[FILLED_SLOTS_CONTENT_CONTEXT_KEY]
-        except KeyError:
-            raise TemplateSyntaxError(f"Attempted to render SlotNode '{self.name}' outside a parent component.")
+        component_id = get_slot_component_association(context, self.node_id)
+        trace_msg("RENDR", "SLOT", self.name, self.node_id, component_id=component_id)
 
+        slot_fill_content = get_slot_fill(context, component_id, self.name)
         extra_context = {}
-        try:
-            slot_fill_content: FillContent = filled_slots_map[(self.name, self.template)]
-        except KeyError:
+
+        # Slot fill was NOT found. Will render the default fill
+        if slot_fill_content is None:
             if self.is_required:
                 raise TemplateSyntaxError(
                     f"Slot '{self.name}' is marked as 'required' (i.e. non-optional), " f"yet no fill is provided. "
                 )
             nodelist = self.nodelist
+
+        # Slot fill WAS found
         else:
             nodelist, alias = slot_fill_content
             if alias:
                 if not alias.isidentifier():
                     raise TemplateSyntaxError()
                 extra_context[alias] = UserSlotVar(self, context)
 
-        with context.update(extra_context):
-            return nodelist.render(context)
-
-
-class BaseFillNode(Node):
-    def __init__(self, nodelist: NodeList):
-        self.nodelist: NodeList = nodelist
-
-    def __repr__(self) -> str:
-        raise NotImplementedError
+        used_ctx = self.resolve_slot_context(context)
+        with used_ctx.update(extra_context):
+            output = nodelist.render(used_ctx)
+
+        trace_msg("RENDR", "SLOT", self.name, self.node_id, component_id=component_id, msg="...Done!")
+        return output
+
+    def resolve_slot_context(self, context: Context) -> Context:
```

Comment:
This `resolve_slot_context` is where we configure the slot context resolution. Basically, this is what the original MR was supposed to be about.